Annotate non public APIs used by the build scan plugin.

felipefzdz opened the pull request:
The scan plugin relies on several internal BT APIs. This is not always clear to anyone working on the BT code, which means that the build scan plugin can unknowingly be broken. We want to make this dependency clear.

felipefzdz commented:
This was requested by @alkemist, so probably it makes sense if it's reviewed by him.

bamboo commented:
> This is not always clear to anyone working on the BT code, which means that the build scan plugin can unknowingly be broken

That's a nice idea and the same is true of APIs used by `gradle-script-kotlin`.

alkemist commented on subprojects/base-services/src/main/java/org/gradle/internal/progress/BuildOperationListener.java:
```diff
@@ -15,11 +15,14 @@
  *
  * @since 3.5
  */
+@UsedByScanPlugin
```
This isn't the case. We used this in Gradle 3.5, but no longer do. As such, it doesn't need this annotation.

alkemist replied:
This isn't the case. We used this in Gradle 3.5, but no longer do. As such, it doesn't need this annotation.

alkemist commented on subprojects/base-services/src/main/java/org/gradle/internal/progress/OperationFinishEvent.java:
```diff
@@ -17,7 +17,9 @@
 import org.gradle.api.Nullable;
+import org.gradle.internal.scan.UsedByScanPlugin;
 
+@UsedByScanPlugin
```
Needs comment that it is used via InternalTaskExecutionListener.

alkemist commented on subprojects/base-services/src/main/java/org/gradle/internal/progress/OperationStartEvent.java:
```diff
@@ -16,6 +16,9 @@
 
+import org.gradle.internal.scan.UsedByScanPlugin;
+
+@UsedByScanPlugin
```
Needs comment that it is used via InternalTaskExecutionListener.

alkemist commented on subprojects/core/src/main/java/org/gradle/api/internal/GradleInternal.java:
```diff
@@ -80,6 +81,7 @@
      */
     BuildListener getBuildListenerBroadcaster();
 
+    @UsedByScanPlugin
```
We should put this on the class as well.

alkemist commented on subprojects/core/src/main/java/org/gradle/api/internal/project/ProjectInternal.java:
```diff
@@ -75,6 +76,7 @@
 
     FileResolver getFileResolver();
 
+    @UsedByScanPlugin
```
On class as well.

alkemist commented on subprojects/core/src/main/java/org/gradle/internal/buildevents/TaskExecutionLogger.java:
```diff
@@ -23,13 +23,15 @@
 /**
  * A listener which logs the execution of tasks.
  */
+@UsedByScanPlugin
```
Needs clarifying comment.

adammurdoch replied:
this one seems iffy to me. what's it used for?

felipefzdz replied:
It's used for not serializing ProgressStartEvent events corresponding to the task execution logger. We're only checking the package and name of the class, but a refactor of that would break our code. I'll add a comment into the annotation.

alkemist commented:
@adammurdoch can you please do a pass through this and drop a note on any dependency that you don't understand. That is, you don't intuitively understand how/why the scan plugin uses it.

felipefzdz commented:
@adammurdoch I've applied more annotations after your approval, as @alkemist let me know that my heuristic to find internal classes was insufficient. I've also addressed your comments.

etiennestuder commented on subprojects/base-services/src/main/java/org/gradle/internal/progress/OperationFinishEvent.java:
```diff
@@ -17,7 +17,9 @@
 import org.gradle.api.Nullable;
+import org.gradle.internal.scan.UsedByScanPlugin;
 
+@UsedByScanPlugin("Used via InternalTaskExecutionListener")
```
@luke @felipefzdz IMO, we should not reference classes that are internal to our build scan plugin in the public source code of Gradle BT.

alkemist replied:
InternalTaskExecutionListener is a Gradle class.

etiennestuder replied:
Missed the `via`. My bad. Thanks for the clarification.

etiennestuder commented on subprojects/base-services/src/main/java/org/gradle/internal/progress/OperationStartEvent.java:
```diff
@@ -16,6 +16,9 @@
 
+import org.gradle.internal.scan.UsedByScanPlugin;
+
+@UsedByScanPlugin("Used via InternalTaskExecutionListener")
```
Same comment as above.
